src/runtime_src/core/tools/xbutil2/OO_P2P.cpp

```diff
@@ -17,12 +17,244 @@
 // Local - Include Files
 #include "OO_P2P.h"
 #include "tools/common/XBUtilities.h"
+#include "core/common/error.h"
+#include "core/common/unistd.h"
+#include "core/common/memalign.h"
+#include "core/common/device.h"
+#include "core/common/query_requests.h"
+#include "experimental/xrt-next.h"
+#include "experimental/xrt_bo.h"
+#include "xrt.h"
+#include "xclbin.h"
 namespace XBU = XBUtilities;
 
 // 3rd Party Library - Include Files
+#include <boost/program_options.hpp>
+#include <boost/algorithm/string/join.hpp>
+namespace po = boost::program_options;
 
 // System - Include Files
 #include <iostream>
+#include <cctype>
+
+#ifdef _WIN32
+# pragma warning(disable : 4245)
+#endif
+
+namespace {
+
+enum class action_type { enable, disable, validate };
+
+action_type
+string2action(std::string str)
+{
+  std::transform(str.begin(), str.end(), str.begin(),
+                 [](auto c) { return static_cast<char>(std::tolower(c)); });
+
+  if (str == "enable")
+    return action_type::enable;
+  else if (str == "disable")
+    return action_type::disable;
+  else if (str == "validate")
+    return action_type::validate;
+  else
+    throw xrt_core::generic_error(EINVAL, "Invalid p2p action '" + str + "'");
+}
+
+static auto
+p2p_config(xrt_core::device* device)
+{
+  try {
+    auto cfg = xrt_core::device_query<xrt_core::query::p2p_config>(device);
+    return cfg;
+  }
+  catch (const std::exception&) {
+    throw xrt_core::system_error(ENOTSUP, "p2p is not supported");
+  }
+}
+
+static void
+p2p_enabled_or_error(xrt_core::device* device)
+{
+  auto cfg = p2p_config(device);
+
+  long long bar = -1;
+  long long rbar = -1;
+  long long remap = -1;
+  long long exp_bar = -1;
+
+  for (auto& str : cfg) {
+    auto pos = str.find(":");
+    auto key = str.substr(0, pos);
+    ++pos; // past ":"
+    if (key == "bar")
+      bar = std::stoll(str.substr(pos));
+    else if (key == "exp_bar")
+      exp_bar = std::stoll(str.substr(pos));
+    else if (key == "rbar")
+      rbar = std::stoll(str.substr(pos));
+    else if (key == "remap")
+      remap = std::stoll(str.substr(pos));
+  }
+
+  if (bar == -1)
+    throw xrt_core::system_error(ENOTSUP,"p2p is not supported");
+  if (rbar != -1 && rbar > bar)
+    throw xrt_core::system_error(EIO,"Please WARM reboot to enable p2p");
+  if (remap > 0 && remap != bar)
+    throw xrt_core::system_error(EIO,"p2p remapper is not set correctly");
+  if (bar == exp_bar)
+    return;
+
+  throw xrt_core::system_error(ENOTSUP,"p2p is not supported");
+}
+
+namespace p2ptest {
+
+static void
+fill_with_stride(char* begin, const char* end, char pattern)
```

Comment:
Could you change the variable "pattern" to something like "fill_byte" or something like that.  I ask for pattern is misleading and the name suggests it could be more then just a fill_byte.

Reply:
Could point.  I was also not liking pattern.